sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala

```diff
@@ -370,4 +368,46 @@ class JsonPartitionReader(
     }
   }
 
+  /**
+   * JSON has strict rules about valid numeric formats. See https://www.json.org/ for specification.
+   *
+   * Spark then has it's own rules for supporting NaN and Infinity, which are not
+   * valid numbers in JSON.
+   */
+  private def sanitizeNumbers(input: ColumnVector): ColumnVector = {
+    // Note that this is not 100% consistent with Spark. See the following issues
+    // for more information:
+    // https://issues.apache.org/jira/browse/SPARK-38060
+    // https://github.com/NVIDIA/spark-rapids/issues/4615
+    val regex = if (parsedOptions.allowNonNumericNumbers) {
+      "^" +
+      "(?:" +
+        "(?:-?[0-9]+(?:\\.[0-9]+)?(?:[eE][\\-\\+]?[0-9]+)?)" +
+        "|NaN" +
+        "|(?:[\\+\\-]?INF)" +
+        "|(?:[\\-\\+]?Infinity)" +
+      ")" +
+      "$"
+    } else {
+      "^-?[0-9]+(?:\\.[0-9]+)?(?:[eE][\\-\\+]?[0-9]+)?$"
```

Comment:
Is there any way we can do this without a regular expression?  If not could we file a follow on issue to figure out how to speed this up?

Reply:
Yes, I can rework this to avoid regexp. I will do so in this PR.

Reply:
I simplified this a bit but it is still calling into `GpuCast` which applies additional complex regular expressions that are mostly redundant because we have already sanitized the input here, so I have filed https://github.com/NVIDIA/spark-rapids/issues/4837 to optimize this farther